beacon-chain/verification/data_column.go

```diff
@@ -0,0 +1,504 @@
+		stateByRoot                 map[[fieldparams.RootLength]byte]state.BeaconState
+	}
+
+	rodataColumnsCommitmentVerifier func([]blocks.RODataColumn) error
+)
+
+var _ DataColumnsVerifier = &RODataColumnsVerifier{}
+
+// VerifiedRODataColumns "upgrades" wrapped RODataColumns to VerifiedRODataColumns.
+// If any of the verifications ran against the data columns failed, or some required verifications
+// were not run, an error will be returned.
+func (dv *RODataColumnsVerifier) VerifiedRODataColumns() ([]blocks.VerifiedRODataColumn, error) {
+	if !dv.results.allSatisfied() {
+		return nil, dv.results.errors(errColumnsInvalid)
+	}
+
+	verifiedRODataColumns := make([]blocks.VerifiedRODataColumn, 0, len(dv.dataColumns))
+	for _, dataColumn := range dv.dataColumns {
+		verifiedRODataColumn := blocks.NewVerifiedRODataColumn(dataColumn)
+		verifiedRODataColumns = append(verifiedRODataColumns, verifiedRODataColumn)
+	}
+
+	return verifiedRODataColumns, nil
+}
+
+// SatisfyRequirement allows the caller to assert that a requirement has been satisfied.
+// This gives us a way to tick the box for a requirement where the usual method would be impractical.
+// For example, when batch syncing, forkchoice is only updated at the end of the batch. So the checks that use
+// forkchoice, like descends from finalized or parent seen, would necessarily fail. Allowing the caller to
+// assert the requirement has been satisfied ensures we have an easy way to audit which piece of code is satisfying
+// a requirement outside of this package.
+func (dv *RODataColumnsVerifier) SatisfyRequirement(req Requirement) {
+	dv.recordResult(req, nil)
+}
+
+func (dv *RODataColumnsVerifier) recordResult(req Requirement, err *error) {
+	if err == nil || *err == nil {
+		dv.results.record(req, nil)
+		return
+	}
+	dv.results.record(req, *err)
+}
+
+func (dv *RODataColumnsVerifier) ValidFields() (err error) {
+	if ok, err := dv.results.cached(RequireValidFields); ok {
+		return err
+	}
+
+	defer dv.recordResult(RequireValidFields, &err)
+
+	for _, dataColumn := range dv.dataColumns {
+		if err := peerdas.VerifyDataColumnSidecar(dataColumn); err != nil {
+			return columnErrBuilder(errors.Wrap(err, "verify data column sidecar"))
+		}
+	}
+
+	return nil
+}
+
+func (dv *RODataColumnsVerifier) CorrectSubnet(expectedTopics []string) (err error) {
+	if ok, err := dv.results.cached(RequireCorrectSubnet); ok {
+		return err
+	}
+
+	defer dv.recordResult(RequireCorrectSubnet, &err)
+
+	if len(expectedTopics) != len(dv.dataColumns) {
+		return columnErrBuilder(errBadTopicLength)
+	}
+
+	for i := range dv.dataColumns {
+		// We add a trailing slash to avoid, for example,
+		// an actual topic /eth2/9dc47cc6/data_column_sidecar_1
+		// to match with /eth2/9dc47cc6/data_column_sidecar_120
+		expectedTopic := expectedTopics[i] + "/"
+
+		actualSubnet := peerdas.ComputeSubnetForDataColumnSidecar(dv.dataColumns[i].Index)
+		actualSubTopic := fmt.Sprintf(dataColumnSidecarSubTopic, actualSubnet)
```

Comment:
This somehow feels different than a lot of the other verification methods. When we wrote the equivalent blob verification code for validating topics, it felt more natural to keep that encapsulated in the sync package. I think I like that you are pushing us further toward consolidating more into the verification package, but as we inch closer to a conceptual circular dependency it starts to seem wrong. I think in this case you needed to copy/paste some subnet construction logic (eg copying`dataColumnSidecarSubTopic` and this Sprintf) so, maybe this is the line where it can just be checked in the gossip handler and excluded from the verifier?

Reply:
Proposal: 59c2e59987e098aaede5670a5ffbbfb7a3f7468b

Reply:
sounds good!